Version 3.0.0

nex3 opened the pull request:
This PR fully implements the changes discussed in #17, integrating this client with [a branch] of @thefifthmatt's static randomizer so that _all_ item locations can be randomized rather than just unique items.

[a branch]: https://github.com/nex3/SoulsRandomizers/tree/archipelago-server

Major functionality changes:

* The static randomizer now replaces all in-world item pickups with synthetic items whose paramdata encodes the Archipelago location ID and (for local items) the DS3 item ID. Specifically:

  * When a local item is picked up, this client decodes the paramdata, tells Archipelago that the corresponding location was accessed, and replaces the synthetic item with the original DS3 item before it enters the player's inventory.
  * When a foreign item is picked up, the client decodes the paramdata, tells Archipelago that the corresponding location was accessed, and deletes the item from the player's inventory.

* Uses ModEngine2 rather than manual dinput8.dll injection to be compatible with other mods (particularly the static randomizer). This also makes it compatible with DS3 1.15.2.

* More gracefully handles connection and disconnection:

   * Shows a banner in-game when a connection is established.
   * If foreign items are picked up while disconnected, they will be sent once reconnected.
   * More robust checks for when the player is fully loaded into the game.

* Drops progressive item logic since that's no longer necessary to handle non-unique items.

* Supports granting Path of the Dragon through a custom event created by the static randomizer.

Major infrastructure changes:

* Uses vcpkg for dependencies where possible (similar to #27).
* Uses MinHook to hook into internal DS3 functionality with AOBs, adding more version flexibility.
* Avoids assembly in favor of explicit C code, manipulating defined structs where possible.
* Adds a GitHub Actions workflow to automatically build and publish tagged releases.
* Uses `spdlog` for compatibility with ModEngine2.

Instructions for building this client are in the README. You can test it by taking [the latest beta release](https://github.com/nex3/Dark-Souls-III-Archipelago-client/releases) and overwriting its `archipelago.dll` with the one you built locally.

Closes #17

nex3 committed:
Get the mod building in both debug and release mode

Also add documentation for getting this building locally.

nex3 committed:
Randomize collaboratively with the offline randomizer

This sets the stage for randomizing to all available locations, including shops.

nex3 committed:
Build directly to archipelago.dll

nex3 committed:
Support newer server structure

nex3 committed:
Fix a few bugs

nex3 committed:
Use a true function call for ItemGib.

This allows us to give multiple items in a much better way.

nex3 committed:
Show banners in-game

nex3 committed:
Add the ability to set events to GameHook

This still needs to be integrated into the online logic.

nex3 committed:
Better support for Path of the Dragon

nex3 committed:
Use an AOB for fSetEventFlag

nex3 committed:
Add an AOB for mapItemMan

nex3 committed:
Use FD4Singleton to check for DLC

nex3 committed:
Use FD4Singleton/WorldChrMan for killThePlayer

nex3 committed:
Use singleton const/destructors to detect when the game is loaded

nex3 committed:
Use WorldChrMan and SprjEventFlagMan for updateRuntimeValues

nex3 committed:
Add AOBs for equip locking

nex3 committed:
Throw errors with names for pattern mismatches

nex3 committed:
Begin the process of making the mod cross-patch

nex3 committed:
Clean up and AOB item gib logic

nex3 committed:
Be more careful about accessing memory during initialization

nex3 committed:
Basic DLL compatibility with Mod Engine 2

nex3 committed:
Use dependencies from vcpkg where possible.

This doesn't use minhook from vcpkg because that version seems to be incompatible somehow, and we want to start using the ModEngine2 hook API anyway.

nex3 committed:
Make this a proper ME2 extension and use its logging infrastructure

nex3 committed:
Rrefactor GameHook.h into a few separate headers

nex3 committed:
Make function hooking a bit more consistent

nex3 committed:
Remove an excess newline

nex3 committed:
Fix AutoEquip, show trace logs in debug mode

nex3 committed:
Use a better technique and UI for showing banner messages

nex3 committed:
Bump version to 3.0.0-beta.1

nex3 committed:
Drop more dinput8 references

nex3 committed:
Remove a leftover debugging print

nex3 committed:
Statically link openssl and don't crash in release mode

nex3 committed:
Bump version

nex3 committed:
Update version

nex3 committed:
Bump version

nex3 committed:
Bump version

nex3 committed:
Remove synthetic items from inventory when connected.

When not connected, keep them around to send when connected.

nex3 committed:
Bump version to 3.0.0-beta.6

nex3 committed:
Build archipelago.dll via GitHub actions on release

nex3 committed:
Update vcpkg version

nex3 committed:
Build the static randomizer as part of the release process

nex3 committed:
Cut a full release when pushing a tag

nex3 committed:
Release 3.0.0-beta.7

nex3 committed:
Fix release YAML

nex3 committed:
Bump version

nex3 committed:
Merge branch 'main' of https://github.com/Marechal-L/Dark-Souls-III-Archipelago-client into full-rando-server

nex3 committed:
Check out YATC as well when building the static rando

Marechal-L commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,177 @@
+    - name: Download Archipelago repo
+      run: |
+        Invoke-WebRequest -OutFile Archipelago.zip -Uri "https://github.com/nex3/Archipelago/archive/refs/tags/ds3-${{ needs.version.outputs.version }}.zip"
+        Expand-Archive Archipelago.zip .
```
Can we remove this from the final version once the server and the client has been merged ?

nex3 replied:
Yes. Even before then this should _really_ only include the apworld, but people have to use my fork of the entire repo until https://github.com/ArchipelagoMW/Archipelago/pull/2403 is merged.

Marechal-L commented on archipelago-client/ArchipelagoInterface.cpp:
```diff
@@ -67,14 +73,21 @@ BOOL CArchipelago::Initialise(std::string URI) {
-		Core->Logger("Slot disconnected");
+		spdlog::info("Slot disconnected");
+		Core->connected = false;
+		GameHook->showBanner(L"Archipelago disconnected! Don't pick up any items until it reconnects");
```
Maybe this behavior belongs also to set_socket_disconnected_handler, I need to give it a try

Marechal-L commented on release/README.txt:
```diff
@@ -0,0 +1,97 @@
+
+In addition, the following options have changed:
+
+* `enable_*_locations` options have all been deprecated. Instead, there are a number of named location groups that you can add to the `exclude_locations` option to prevent them from containing important items. These location groups are:
```
Just like in the AP server, I suggest to mark those locations as removed while keeping the explainations on how to setup "exclude_locations" properly

Marechal-L commented:
I don't encountered any major issues during my testings and the result looks amazing! 

I still have two minor issues and I would be glad to investigate on my own to fix this but I would like your feedback first : 

1. Items injected by the SoulsRandomizers are displayed with the default label "?good's name?" when the game is played in French ( and probably any language but English ). I didn't dig into the SoulsRandomizers yet but if you have any idea to fix this it would be great.

2. Local items purchased from the shop do not trigger a check to the server but world items do ? I'll probably need to do more testing to fully understand your logic behind it but if you already have an explanation for me I would be greateful

nex3 committed:
Code review

nex3 commented:
> Items injected by the SoulsRandomizers are displayed with the default label "?good's name?" when the game is played in French ( and probably any language but English ). I didn't dig into the SoulsRandomizers yet but if you have any idea to fix this it would be great.

This is currently a restriction intrinsic to the static randomizer—it only overwrites the messages for the `engus` localization. Hopefully we can talk to Matt about making that more flexible, but I think it may have to be a post-3.0.0 thing.

> Local items purchased from the shop do not trigger a check to the server but world items do ? I'll probably need to do more testing to fully understand your logic behind it but if you already have an explanation for me I would be greateful

Yeah, this is definitely not ideal but it's _extremely_ hard to fix. The core of the issue is, in order to make DS3-native shop items look right (show up in the right tabs, in the right order, with the right stats) we need them to be _real_ items rather than the synthetic fakes we use for other drops. But since they're real items, we can't embed extra metadata to figure out which location they represent once you purchase them.

It's possible we could work around this at least to some degree if we could get a listing of all the items that exist in the shop and use that to determine which location was bought... although even then if you have, for example, six different Embers available in the Firelink shop from different locations it's not clear how you'd figure out which Ember came from which source.

nex3 committed:
Update release process

* Presets, scripts, and sfx are now stored directly in the static
  randomizer repo and don't need to be pulled forward from an older
  release.

* We can now distribute the Archipelago logic as a standalone apworld

nex3 committed:
Release 3.0.0-beta.10

nex3 committed:
Merge branch 'full-rando-server' of https://github.com/nex3/Dark-Souls-III-Archipelago-client into full-rando-server

nex3 committed:
Update vcpkg baseline

nex3 committed:
Get this building with the latest visual studio

nex3 committed:
Debug and maybe fix some msbuild/vcpkg stuff

nex3 committed:
Try using an older builtin-baseline

nex3 committed:
Explain apworld usage in more detail

nex3 committed:
Use the non-deprecated openssl package

nex3 committed:
Try building with an explicit registry checkout

nex3 committed:
Bump version to 3.0.0-beta.11

nex3 committed:
Proeprly auto-equip foreign items

Thanks to @eldsmith on the Archipelago discord for this fix.

eldsmith committed:
Auto equip fix, handles shields and legs fix (#1)

Shield support fixes crashes on local check pickups

nex3 committed:
Release 3.0.0-beta.12

nex3 committed:
Support reporting local shop items

nex3 committed:
Merge branch 'full-rando-server' of https://github.com/nex3/Dark-Souls-III-Archipelago-client into full-rando-server

nex3 committed:
Release 3.0.0-beta.13

nex3 committed:
Remove leftover debugging print

nex3 committed:
Automate more of the release process + release ds3-stable

nex3 committed:
Release 3.0.0-beta.14

nex3 committed:
Avoid an unnecessary checkout

nex3 committed:
Fix a merge

nex3 committed:
Don't pull from a tag that doesn't exist yet

nex3 committed:
Try a full clone

nex3 committed:
Fix merge again

nex3 committed:
Add a GitHub robot user for the merge commits

nex3 committed:
More fixes

nex3 committed:
Run static tags in the correct working directory

nex3 committed:
Use the correct directory name

nex3 committed:
Release 3.0.0-beta.15

nex3 committed:
Build the DS3 YAML against 0.5.0

nex3 committed:
Release 3.0.0-beta.16

nex3 committed:
Release 3.0.0-beta.17

nex3 committed:
Only create tags after the release completes

nex3 committed:
Update the release process now that the 3.0.0 PR has landed

nex3 committed:
Release 3.0.0-beta.18

nex3 committed:
Add missing runs-on fields

nex3 committed:
Fix the Archipelago source zip URL

nex3 committed:
Report local PotD to the server

nex3 committed:
Merge branch 'full-rando-server' of https://github.com/nex3/Dark-Souls-III-Archipelago-client into full-rando-server

nex3 committed:
Release 3.0.0-beta.19

nex3 committed:
Add a missing GitHub Actions dependency

nex3 committed:
Release 3.0.0-beta.20

nex3 committed:
Release a self-contained .NET app as well

nex3 committed:
Release 3.0.0-beta.21

nex3 committed:
Mark releases as pre-release

nex3 committed:
Fix SoulsRandomizers publish paths

nex3 committed:
Fix and minimize publish paths

nex3 committed:
Add missing quotes

nex3 committed:
Download and rename proper self-contained executable

nex3 committed:
Remove rename

Turns out this downloads with the original filename, not the artifact name! Weird.

nex3 committed:
Document self contained zip file

nex3 committed:
Release 3.0.0-beta.22

nex3 committed:
Release 3.0.0-beta.23

nex3 committed:
Drop the self-contained release

This doesn't actually run on Proton, which was the primary motivation for providing it, so it's just a waste of resources and confusing for users.

nex3 committed:
Release 3.0.0-beta.24

Commit:
Update spdlog to v1.14.1

Samonella7 committed:
Idiot-proof the "Building Locally" instructions

Samonella7 committed:
Typo fix in "Building Locally" instructions

nex3 committed:
Merge pull request #2 from Samonella7/build-improvements

Improve build instructions

nex3 committed:
Release 3.0.0-beta.25

nex3 committed:
Avoid needing to load MINT

This seems to have stopped compiling, so this relies on a fork of the now-defunct ME2 to avoid loading it.

nex3 committed:
Release 3.0.0-beta.26

nex3 committed:
Release 3.0.0-beta.27

nex3 committed:
Update to new version of ME2 fork
